Multiple definition of `Crate` and `Module` in `libCamShr.so` (GCC 10 -fno-common)

Hi All,

I'm trying to compile MDSplus from source (trying to make an AUR package on arch linux). Compilation goes mostly fine with these commands:

```bash
autoreconf -fi -I m4
./configure
make
```

but during make a code issue pops up
(on stable version 7.96.14-7.9.11 at least)

```
gcc -Wall -Wextra  -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -fvisibility=hidden  -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -fpic -shared-libgcc -fsigned-char -fno-strict-aliasing  -I/home/daan/Projects/AUR/mdsplus/src/MDSplus/include -I/home/daan/Projects/AUR/mdsplus/src/MDSplus/include -D_FORTIFY_SOURCE=2 -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -m64 -shared  -nostartfiles -Wl,-soname=libCamShr.so -o /home/daan/Projects/AUR/mdsplus/src/MDSplus/lib64/libCamShr.so.1.8 add_entry.o bisearch.o cam_functions.o check_for_file.o check_sema4.o commit_entry.o compare_str.o copy.o create_sema4.o create_tmp_file.o debug.o expand_db.o get_crate_status.o get_db_file_size.o get_file_count.o get_file_name.o get_scsi_device_number.o global.o issort.o lock_file.o lookup_entry.o map_data_file.o map_scsi_device.o msg.o parse_crate_db.o parse_cts_db.o QueryHighwayType.o remove_entry.o scsi_io.o ScsiSystemStatus.o turn_crate_on_off_line.o unlock_file.o xlate_logicalname.o RemCamMulti.o RemCamSingle.o RemCamIosb.o -L/home/daan/Projects/AUR/mdsplus/src/MDSplus/lib64/ -lMdsShr -lMdsIpShr 
/bin/ld: bisearch.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: bisearch.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: cam_functions.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: cam_functions.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: check_for_file.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: check_for_file.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: check_sema4.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: check_sema4.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: commit_entry.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: commit_entry.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: compare_str.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: compare_str.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: copy.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: copy.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: create_sema4.o:(.bss+0x10): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: create_sema4.o:(.bss+0x40): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: create_tmp_file.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: create_tmp_file.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: debug.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: debug.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: expand_db.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: expand_db.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: get_crate_status.o:(.bss+0x0): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: get_crate_status.o:(.bss+0x60): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: get_db_file_size.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: get_db_file_size.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: get_file_count.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: get_file_count.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: get_file_name.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: get_file_name.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: get_scsi_device_number.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: get_scsi_device_number.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: lock_file.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: lock_file.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: lookup_entry.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: lookup_entry.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: map_data_file.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: map_data_file.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: map_scsi_device.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: map_scsi_device.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: parse_crate_db.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: parse_crate_db.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: parse_cts_db.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: parse_cts_db.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: QueryHighwayType.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: QueryHighwayType.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: remove_entry.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: remove_entry.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: ScsiSystemStatus.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: ScsiSystemStatus.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: turn_crate_on_off_line.o:(.bss+0x0): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: turn_crate_on_off_line.o:(.bss+0x60): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: unlock_file.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: unlock_file.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
/bin/ld: xlate_logicalname.o:(.bss+0x0): multiple definition of `Crate'; add_entry.o:(.bss+0x0): first defined here
/bin/ld: xlate_logicalname.o:(.bss+0x20): multiple definition of `Module'; add_entry.o:(.bss+0x20): first defined here
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:101: /home/daan/Projects/AUR/mdsplus/src/MDSplus/lib64/libCamShr.so] Error 1
```
Can someone comment?
Perhaps this is a known issue which is circumvented through build flags?

--- a/camshr/crate.h
+++ b/camshr/crate.h
@@ -52,7 +52,7 @@ struct CRATE {
 
 //-------------------------------------------------------------------------
 // internal structure
-struct Crate_ {
+typedef struct Crate_ {
   char name[7];
   int device;
   int type;

--- a/camshr/module.h
+++ b/camshr/module.h
@@ -39,7 +39,7 @@ struct MODULE {
 #define	MODULE_ENTRY	sizeof(struct MODULE)
 
 // internal structure
-struct Module_ {
+typedef struct Module_ {
   char name[32];		// 2002.02.07
   int adapter;
   int id;

--- a/camshr/verbs.c
+++ b/camshr/verbs.c
@@ -290,7 +290,7 @@ EXPORT int Autoconfig(void *ctx __attribute__ ((unused)), char **error, char **o
 //-------------------------------------------------------------------------
 EXPORT int Deassign(void *ctx, char **error, char **output __attribute__ ((unused)))
 {
-  char db_tmp[64];
+  char db_tmp[128];
   int i, modulesToDeassign, modulesDeassigned, numOfEntries, physical_name;
   int status = SUCCESS;
   struct Module_ Mod, *pMod;
@@ -531,7 +531,7 @@ EXPORT int ShowCrate(void *ctx, char **error, char **output)
 //-------------------------------------------------------------------------
 EXPORT int ShowModule(void *ctx, char **error, char **output)
 {
-  char db_tmp[64];		// enough space for a logical name and a cstring terminator     [2002.02.20]
+  char db_tmp[128];		// enough space for a logical name and a cstring terminator     [2002.02.20]
   int i, numOfEntries, status = SUCCESS;
 
 

--- a/deploy/os/bootstrap.opts
+++ b/deploy/os/bootstrap.opts
@@ -1,1 +1,1 @@
---platform=bootstrap --dockerimage=mdsplus/builder:debian_stretch-64 --release
+--platform=bootstrap --dockerimage=mdsplus/builder:bootstrap --release

--- a/deploy/os/fc32.env
+++ b/deploy/os/fc32.env
@@ -1,0 +1,1 @@
+export FFLAGS='-fallow-argument-mismatch -fallow-invalid-boz'

--- a/deploy/packaging/linux-rhel8.xml
+++ b/deploy/packaging/linux-rhel8.xml
@@ -25,7 +25,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 -->
 <deploy>
  <packages>
-  <package name="MDSplus" arch="bin" summary="MDSplus Data Acquisition System" description="MDSplus Data Acquisition System">
+  <package name="MDSplus" arch="noarch" summary="MDSplus Data Acquisition System" description="MDSplus Data Acquisition System">
     <requires package="kernel"/>
     <requires package="java"/>
     <requires package="python"/>

--- a/deploy/packaging/linux.xml
+++ b/deploy/packaging/linux.xml
@@ -25,7 +25,7 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 -->
 <deploy>
  <packages>
-  <package name="MDSplus" arch="bin" summary="MDSplus Data Acquisition System" description="MDSplus Data Acquisition System">
+  <package name="MDSplus" arch="noarch" summary="MDSplus Data Acquisition System" description="MDSplus Data Acquisition System">
     <requires package="kernel"/>
     <requires package="java"/>
     <requires package="python"/>

--- a/include/xmdsshr.h
+++ b/include/xmdsshr.h
@@ -56,5 +56,5 @@ extern EXPORT void XmdsUsageIconsInitialize(Widget w);
 extern EXPORT Pixmap *XmdsUsageIcons();
 extern EXPORT Pixmap *XmdsUsageGrayIcons();
 
-EXPORT WidgetClass xmdsWaveformWidgetClass;
-EXPORT XmdsWaveformClassRec xmdsWaveformClassRec;
+extern EXPORT WidgetClass xmdsWaveformWidgetClass;
+extern EXPORT XmdsWaveformClassRec xmdsWaveformClassRec;

--- a/math/Makefile.in
+++ b/math/Makefile.in
@@ -7,6 +7,12 @@ VPATH=@srcdir@
 #### Remove following line after code causing warnings is fixed ####
 FCFLAGS+=-Wno-compare-reals -Wno-uninitialized
 
+# WARNING: gfortran rises a huge amount of Warning that are converted to errors in deploy script
+#          we had to add -Wno-error to suppress errors in fortran here ...
+FFLAGS += $(FCFLAGS) -Wno-error
+
+LIBPREFIX=MdsMath
+
 SOURCES = \
 	caxpy.f \
 	cgeco.f \
@@ -51,11 +57,9 @@ OBJECTS = $(SOURCES:.f=.o) boxsmooth.o
 
 ## suppressing conversion error warning ##
 umach.o : umach.f
-	$(COMPILE.f) $< $(OUTPUT_OPTION) -Wno-conversion
+	$(COMPILE.f) $< $(OUTPUT_OPTION) $(FFLAGS)
 
 local@LIBPRE@MdsMath = @LIBPRE@MdsMath@SHARETYPE@
-
-FFLAGS += $(FCFLAGS)
 
 ifeq "@MACOSX@" ""
 LIBRARIES = $(@LIBPRE@MdsMath)

--- a/mdslib/Makefile.in
+++ b/mdslib/Makefile.in
@@ -1,5 +1,7 @@
 include @top_builddir@/Makefile.inc
 
+LIBPREFIX=MdsLib
+
 srcdir=@srcdir@
 builddir=@builddir@ 
 VPATH=@srcdir@
@@ -11,6 +13,10 @@ MKDIR_P=@MKDIR_P@
 
 LIBS = @LIBS@ @GLOBUS_LIBS@ $(LIBSOCKET) @LIBRESOLV@ @LIBM@
 MDSLIB_LIBS=-lTdiShr -lTreeShr -lMdsShr
+
+# WARNING: gfortran rises a huge amount of Warning that are converted to errors in deploy script
+#          we had to add -Wno-error to suppress errors in fortran here ...
+FFLAGS += $(FCFLAGS) -Wno-error
 
 SOURCES = MdsLib.c
 OBJECTS = $(SOURCES:.c=.o)

--- a/mitdevices/paragon_hist.c
+++ b/mitdevices/paragon_hist.c
@@ -154,7 +154,7 @@ static int Store(struct descriptor *niddsc_ptr __attribute__ ((unused)), InStore
   file = fopen(rpt_name.pointer, "r");
   if (file != NULL) {
     while (fgets(line, 255, file)) {
-      TIME time;
+      TIME time = {0, 0};
       float value;
       if (strlen(line) >= 70) {
 	if (ParseHistorian(line, (struct descriptor *)&name, &value, &time)) {

--- a/python/MDSplus/descriptor.py
+++ b/python/MDSplus/descriptor.py
@@ -88,7 +88,7 @@ class Descriptor(object):
             exec(compile("struct.%s = v"%k,'<string>','exec'))
 
     def __new__(cls,obj_in=None,_dict_={}):
-        if cls is not Descriptor or not obj_in:
+        if cls != Descriptor or not obj_in:
             return object.__new__(cls)
         if not obj_in and not hasattr(cls,'__del__'):
             return DescriptorNULL
@@ -119,11 +119,11 @@ class Descriptor(object):
         self.ptr_= _C.cast(self.ptr,Descriptor.PTR)
 
     def __getattr__(self,name):
-        if name is not '_structure' and name in dict(self._structure._fields_):
+        if name != '_structure' and name in dict(self._structure._fields_):
             return self._structure.__getattribute__(name)
         return super(Descriptor,self).__getattr__(name)
     def __setattr__(self,name,value):
-        if name is not '_structure' and name in dict(self._structure._fields_):
+        if name != '_structure' and name in dict(self._structure._fields_):
             return self._structure.__setattr__(name,value)
         return super(Descriptor,self).__setattr__(name,value)
 

